flink-connector-gcp-pubsub/src/main/java/org/apache/flink/connector/gcp/pubsub/sink/PubSubWriter.java

```diff
@@ -0,0 +1,195 @@
+        Preconditions.checkNotNull(context, "Context cannot be null.");
+
+        this.mailboxExecutor = context.getMailboxExecutor();
+        this.numBytesOutCounter = context.metricGroup().getIOMetricGroup().getNumBytesOutCounter();
+        this.numRecordsOutCounter =
+                context.metricGroup().getIOMetricGroup().getNumRecordsOutCounter();
+        this.numRecordsOutErrorCounter = context.metricGroup().getNumRecordsOutErrorsCounter();
+        this.maximumInflightRequests = maximumInflightRequests;
+        this.failOnError = failOnError;
+    }
+
+    @Override
+    public void write(T t, SinkWriter.Context context) throws IOException, InterruptedException {
+        awaitMaxInflightRequestsBelow(maximumInflightRequests);
+        PubsubMessage message =
+                PubsubMessage.newBuilder()
+                        .setData(ByteString.copyFrom(serializationSchema.serialize(t)))
+                        .build();
+        publishMessage(message);
+    }
+
+    @Override
+    public void flush(boolean b) throws IOException, InterruptedException {
+        publisher.flush();
+        awaitMaxInflightRequestsBelow(1);
+    }
+
+    private void awaitMaxInflightRequestsBelow(long maxInflightRequests)
+            throws InterruptedException {
+        while (inflightRequests >= maxInflightRequests) {
+            mailboxExecutor.yield();
+        }
+    }
+
+    private void publishMessage(PubsubMessage message) {
+        ApiFuture<String> future = publisher.publish(message);
+        inflightRequests++;
+        ApiFutures.addCallback(
+                future,
+                new ApiFutureCallback<String>() {
+                    @Override
+                    public void onFailure(Throwable throwable) {
+                        mailboxExecutor.execute(
+                                () -> failMessage(message, throwable),
+                                "Handling failure of Pub/Sub message publishing.");
+                    }
+
+                    @Override
+                    public void onSuccess(String ignored) {
+                        mailboxExecutor.execute(
+                                () -> completeMessage(message),
+                                "Handling success of Pub/Sub message publishing.");
+                    }
+                },
+                directExecutor());
+    }
+
+    private void completeMessage(PubsubMessage message) {
+        inflightRequests--;
+        numBytesOutCounter.inc(message.getSerializedSize());
+        numRecordsOutCounter.inc();
+    }
+
+    private void failMessage(PubsubMessage message, Throwable throwable) {
+        inflightRequests--;
+        numRecordsOutErrorCounter.inc();
+        raiseIfFatal(throwable);
+
+        if (failOnError) {
+            throw new FlinkRuntimeException("Failed to publish message to Pub/Sub", throwable);
+        }
+
+        publishMessage(message);
```

Comment:
Maybe add the number of retries on failure as a configuration and default value?

Reply:
@morazow I think this is a good idea and also possibly timeout for requests since these can cause deadlocks on flush similar to what we are trying to fix for [AsyncSink](https://cwiki.apache.org/confluence/display/FLINK/FLIP-451%3A+Introduce+timeout+configuration+to+AsyncSink+API), what do you think about following up with another PR to not inflate this one?
cc @snuyanzin